stdlib/public/runtime/SwiftObject.mm

```diff
@@ -373,11 +374,62 @@ + (BOOL)conformsToProtocol:(Protocol*)proto {
 }
 
 - (NSUInteger)hash {
-  return (NSUInteger)self;
+  auto selfMetadata = _swift_getClassOfAllocated(self);
+  auto hashableConformance =
+    reinterpret_cast<const hashable_support::HashableWitnessTable *>(
+      swift_conformsToProtocolCommon(
+	selfMetadata, &hashable_support::HashableProtocolDescriptor));
+  if (hashableConformance == NULL) {
+    return (NSUInteger)self;
+  }
+  return _swift_stdlib_Hashable_hashValue_indirect(
+    &self, selfMetadata, hashableConformance);
 }
 
-- (BOOL)isEqual:(id)object {
-  return self == object;
+- (BOOL)isEqual:(id)other {
+  if (self == other) {
+    return YES;
+  }
+  // Both objects must be `SwiftObject` in ObjC
+  if (object_getClass(other) != object_getClass((id)self)) {
```

Comment:
This is incorrect in the face of subclassing. If we get here, we have to go into the following code to search for a common parent. However, the classes will be the same in the common case, so short-circuiting the common-parent code when the classes are equal would be productive.

Reply:
The intent here is to make sure that `other` really came from Swift.

My understanding is that there are three different types for each object here:
* The type that's in the Obj-C header in memory -- I thought this was always _SwiftObject
* The Obj-C type returned from `-class` -- this is dynamically looked up in _SwiftObject's -class implementation
* The Swift type of the object

I thought that `object_getClass()` always returned the first of these.

Reply:
Ah, not quite. The root class is `SwiftObject`, but the instance's class is the Swift class. Swift classes are ObjC classes. So if you have:

```
class C {}
object_getClass(C())
```

This will return `C`.

There's one case where things differ a little bit, which is pure ObjC classes. This is because pure ObjC classes are not valid Swift metadata, so they get a wrapper on the Swift end of things:

```
object_getClass(NSObject()) // returns the NSObject class
type(of: NSObject()) // returns a Swift wrapper metadata that points to NSObject
```

But that's the only case where things are like that.

Reply:
Thanks for that clarification.  I believe I have that sorted out now.